docs/proposals/network-policy.md

```diff
@@ -0,0 +1,300 @@
+	// Currently this field only supports "DefaultDeny", but could 
+	// be extended to support other policies in the future.  When set to DefaultDeny,
+	// pods in this namespace are denied ingress traffic.  When not defined,
+	// the cluster default ingress policy is applied (currently allow all). 
+	Isolation *IngressIsolationPolicy `json:"isolation,omitempty"` 
+}
+```
+
+```yaml
+kind: Namespace
+apiVersion: v1
+spec:
+  networkPolicy:
+    ingress:
+      isolation: DefaultDeny
+```
+
+The above structures will be represented in v1beta as a json encoded annotation like so:
+
+```yaml
+kind: Namespace
+apiVersion: v1
+metadata:
+  annotations:
+    net.beta.kubernetes.io/ingress-isolation: DefaultDeny 
+```
+
+### NetworkPolicy Go Definition
+
+Once a Namespace is isolated, a mechanism to selectively allow traffic into the namespace and between pods within
+the namespace is required.  That is accomplished through ingress rules on `NetworkPolicy`
+objects (of which there can be multiple in a single namespace).  Pods selected by
+one or more NetworkPolicy objects should allow any incoming connections that match any
+ingress rule on those NetworkPolicy objects, per the network plugin’s capabilities.
+
+If `ingress-isolation` is not specified on a namespace, then all traffic is allowed to pods in that namespace.
+
+```go
+type NetworkPolicy struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior for this NetworkPolicy.
+  Spec NetworkPolicySpec 
+}
+
+type NetworkPolicySpec struct {
+  // Selects the pods to which this NetworkPolicy object
+  // applies.  The array of NetworkPolicyIngressRules below 
+  // is applied to any pods selected by this field.
+  // Multiple NetworkPolicy objects can select the same set of pods.  In this case,
+  // the NetworkPolicyRules for each are combined additively.
+  PodSelector *unversioned.LabelSelector `json:"podSelector,omitempty"`
+
+  // List of ingress rules to be applied to the selected pods.
+  // Traffic is allowed to a pod if Namespace.NetworkPolicy.Ingress.Isolation is undefined, 
+  // OR if the traffic source is the pod's host (for kubelet health checks), 
+  // OR if Namespace.NetworkPolicy.Ingress.Isolation=DefaultDeny and the traffic matches at least 
+  // one NetworkPolicyIngressRule across all of the NetworkPolicy 
+  // objects whose podSelector matches the pod.  If this field is 
+  // empty, this NetworkPolicy has no effect on selected pods.
+  Ingress []NetworkPolicyIngressRule `json:"ingress,omitempty"`
+}
+
+type NetworkPolicyIngressRule struct {
+  // List of ports which should be made accessible on the pods selected by PodSelector.  
+  // Each item in this list is combined using a logical OR.  If this field
+  // is empty, then traffic should not be restricted based on port.  If this 
+  // field is not empty, traffic must match at least one NetworkPolicyPort in the list
+  // or else it will be dropped.
```

Comment:
"or else it will be dropped" is not accurate, the "or else" depends on what other NetworkPolicyIngressRule and NetworkPolicy objects exist.  A more accurate statement might be something like "If this field is not empty then this NetworkPolicyIngressRule grants access only to traffic matching at least one member of this (the Ports) list".